In [1]:
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np
import ast

# Data exploration

## Khám phá dữ liệu về đội hình trong tập dữ liệu

### Đọc data từ file

In [2]:
data_team_df = pd.read_csv('process_data_tft_general_csv.csv', usecols=['Puuid', 'ID Match', 'Placement', 'List Augments', 'List Traits', 'List Champions', 'List Number Items', 'List Champions Tier'])

In [3]:
data_team_df[:10]

Puuid       ID Match  \
0  _NjJNkxpeKQ-po15A8N1GVKYqYTAjiN1WoqkVUGfwmpHqg...  KR_6267320509   
1  WgAZ0CQHGmDf8yctPx5SXkU-rt-9HWv758OmfE6t_Wod8X...  KR_6267320509   
2  WO7vH445Mvh-59sDpk4OAtffIxLMrm2R4wsrGT5ZcUb7cN...  KR_6267320509   
3  LjGpzsuqxxKX_8HDCyOmX5bBgwC-p59Oj7rTLhaSwOgN7U...  KR_6267320509   
4  XdejWcFYkQS2jJwSPR2FXlv_IZuw1txCPOLZJtUkqyusTn...  KR_6267320509   
5  CtB6YjZTon2Z-pYgNfubDLUKwxAZoBQcZVQ12LyNanULi3...  KR_6267320509   
6  y6gI4h6o614IhVvGBBp45jA-XhRHxqx2tleVdE0egluOSK...  KR_6267320509   
7  YFHBmecEZcAKDUkoNKCvM_--z5qekvafrDJzBgqJFz6l5a...  KR_6267320509   
8  y6gI4h6o614IhVvGBBp45jA-XhRHxqx2tleVdE0egluOSK...  KR_6267390363   
9  GSXdu-IDqnhxAjE9OC-rEbktUy7Q-YfRmdz7P6nztlNYHj...  KR_6267390363   

   Placement                                      List Augments  \
0          1  ['TFT7_Augment_Preparation', 'TFT8_Augment_ViC...   
1          4  ['TFT8_Augment_ChannelerTrait', 'TFT8_Augment_...   
2          7  ['TFT6_Augment_ItemGrabBag1', 'TFT8_Augment_Yu...   
3          2  ['TFT6_Augment_SecondWind1', 'TFT8_Augment_Yas...   
4          5  ['TFT6_Augment_ItemGrabBag1', 'TFT8_Augment_Yu...   
5          6  ['TFT6_Augment_TinyTitans', 'TFT8_Augment_Cami...   
6          3  ['TFT6_Augment_TinyTitans', 'TFT8_Augment_Sivi...   
7          8  ['TFT6_Augment_ItemGrabBag1', 'TFT8_Augment_An...   
8          6  ['TFT8_Augment_AnnieSupport', 'TFT6_Augment_Se...   
9          7  ['TFT8_Augment_RellCarry', 'TFT6_Augment_Luden...   

                                         List Traits  \
0  ['Set8_Aegis', 1, 0, 'Set8_Brawler', 2, 1, 'Se...   
1  ['Set8_Aegis', 2, 1, 'Set8_Channeler', 5, 2, '...   
2  ['Set8_Aegis', 1, 0, 'Set8_AnimaSquad', 1, 0, ...   
3  ['Set8_Admin', 1, 0, 'Set8_AnimaSquad', 1, 0, ...   
4  ['Set8_Admin', 1, 0, 'Set8_Aegis', 2, 1, 'Set8...   
5  ['Set8_Admin', 4, 2, 'Set8_Aegis', 1, 0, 'Set8...   
6  ['Set8_Ace', 1, 1, 'Set8_Civilian', 1, 1, 'Set...   
7  ['Set8_Admin', 1, 0, 'Set8_Aegis', 1, 0, 'Set8...   
8  ['Set8_Admin', 1, 0, 'Set8_Aegis', 2, 1, 'Set8...   
9  ['Set8_Aegis', 2, 1, 'Set8_Channeler', 2, 1, '...   

                                      List Champions  \
0  ['TFT8_Gangplank', 'TFT8_Vi', 'TFT8_LeeSin', '...   
1  ['TFT8_Lux', 'TFT8_Annie', 'TFT8_Yuumi', 'TFT8...   
2  ['TFT8_Galio', 'TFT8_Nasus', 'TFT8_Gangplank',...   
3  ['TFT8_Fiora', 'TFT8_Yasuo', 'TFT8_Vayne', 'TF...   
4  ['TFT8_Lulu', 'TFT8_Annie', 'TFT8_Yuumi', 'TFT...   
5  ['TFT8_Talon', 'TFT8_Blitzcrank', 'TFT8_Vi', '...   
6  ['TFT8_Sivir', 'TFT8_Velkoz', 'TFT8_Chogath', ...   
7  ['TFT8_Lulu', 'TFT8_Annie', 'TFT8_Yuumi', 'TFT...   
8  ['TFT8_Lulu', 'TFT8_Annie', 'TFT8_Sona', 'TFT8...   
9  ['TFT8_Rell', 'TFT8_Yuumi', 'TFT8_Nilah', 'TFT...   

          List Number Items       List Champions Tier  
0     [1, 3, 0, 2, 3, 3, 0]     [2, 3, 2, 3, 3, 3, 1]  
1  [1, 3, 1, 0, 0, 1, 3, 0]  [2, 3, 2, 2, 2, 2, 2, 1]  
2  [1, 0, 3, 0, 0, 3, 0, 1]  [2, 2, 3, 2, 2, 3, 2, 1]  
3  [0, 1, 1, 0, 1, 1, 3, 3]  [2, 2, 2, 2, 2, 2, 2, 2]  
4  [0, 3, 1, 0, 1, 3, 1, 3]  [2, 2, 2, 2, 2, 2, 2, 2]  
5  [0, 2, 0, 3, 0, 0, 1, 1]  [2, 2, 1, 3, 1, 2, 2, 2]  
6  [0, 0, 3, 0, 3, 3, 0, 1]  [2, 2, 3, 2, 2, 2, 2, 1]  
7  [0, 3, 1, 0, 3, 3, 1, 0]  [2, 2, 1, 2, 2, 1, 1, 1]  
8     [0, 3, 0, 0, 3, 1, 1]     [2, 3, 2, 1, 2, 2, 2]  
9  [3, 0, 0, 1, 0, 1, 3, 0]  [2, 2, 2, 1, 1, 2, 1, 1]

## Data exploration

Dữ liệu có bao nhiêu cột và hàng:

In [4]:
row, col = data_team_df.shape
row, col

(32000, 8)

Kiểm tra dữ liệu trùng:

In [5]:
have_duplicated_rows = data_team_df.duplicated().sum()
have_duplicated_rows

13904

Tổng cổng có 13904 đội hình bị trùng nhau trên tổng số 32000 đội hình. Sở dĩ con số lớn như vậy là do các người chơi có thể chơi cùng nhau rất nhiều trận nên việc lặp là khá nhiều.

### Loại bỏ các đội hình bị trùng trong các trận đấu:

In [6]:
data_team_df = data_team_df.drop_duplicates().reset_index(drop=True)
data_team_df[:10]

Puuid       ID Match  \
0  _NjJNkxpeKQ-po15A8N1GVKYqYTAjiN1WoqkVUGfwmpHqg...  KR_6267320509   
1  WgAZ0CQHGmDf8yctPx5SXkU-rt-9HWv758OmfE6t_Wod8X...  KR_6267320509   
2  WO7vH445Mvh-59sDpk4OAtffIxLMrm2R4wsrGT5ZcUb7cN...  KR_6267320509   
3  LjGpzsuqxxKX_8HDCyOmX5bBgwC-p59Oj7rTLhaSwOgN7U...  KR_6267320509   
4  XdejWcFYkQS2jJwSPR2FXlv_IZuw1txCPOLZJtUkqyusTn...  KR_6267320509   
5  CtB6YjZTon2Z-pYgNfubDLUKwxAZoBQcZVQ12LyNanULi3...  KR_6267320509   
6  y6gI4h6o614IhVvGBBp45jA-XhRHxqx2tleVdE0egluOSK...  KR_6267320509   
7  YFHBmecEZcAKDUkoNKCvM_--z5qekvafrDJzBgqJFz6l5a...  KR_6267320509   
8  y6gI4h6o614IhVvGBBp45jA-XhRHxqx2tleVdE0egluOSK...  KR_6267390363   
9  GSXdu-IDqnhxAjE9OC-rEbktUy7Q-YfRmdz7P6nztlNYHj...  KR_6267390363   

   Placement                                      List Augments  \
0          1  ['TFT7_Augment_Preparation', 'TFT8_Augment_ViC...   
1          4  ['TFT8_Augment_ChannelerTrait', 'TFT8_Augment_...   
2          7  ['TFT6_Augment_ItemGrabBag1', 'TFT8_Augment_Yu...   
3          2  ['TFT6_Augment_SecondWind1', 'TFT8_Augment_Yas...   
4          5  ['TFT6_Augment_ItemGrabBag1', 'TFT8_Augment_Yu...   
5          6  ['TFT6_Augment_TinyTitans', 'TFT8_Augment_Cami...   
6          3  ['TFT6_Augment_TinyTitans', 'TFT8_Augment_Sivi...   
7          8  ['TFT6_Augment_ItemGrabBag1', 'TFT8_Augment_An...   
8          6  ['TFT8_Augment_AnnieSupport', 'TFT6_Augment_Se...   
9          7  ['TFT8_Augment_RellCarry', 'TFT6_Augment_Luden...   

                                         List Traits  \
0  ['Set8_Aegis', 1, 0, 'Set8_Brawler', 2, 1, 'Se...   
1  ['Set8_Aegis', 2, 1, 'Set8_Channeler', 5, 2, '...   
2  ['Set8_Aegis', 1, 0, 'Set8_AnimaSquad', 1, 0, ...   
3  ['Set8_Admin', 1, 0, 'Set8_AnimaSquad', 1, 0, ...   
4  ['Set8_Admin', 1, 0, 'Set8_Aegis', 2, 1, 'Set8...   
5  ['Set8_Admin', 4, 2, 'Set8_Aegis', 1, 0, 'Set8...   
6  ['Set8_Ace', 1, 1, 'Set8_Civilian', 1, 1, 'Set...   
7  ['Set8_Admin', 1, 0, 'Set8_Aegis', 1, 0, 'Set8...   
8  ['Set8_Admin', 1, 0, 'Set8_Aegis', 2, 1, 'Set8...   
9  ['Set8_Aegis', 2, 1, 'Set8_Channeler', 2, 1, '...   

                                      List Champions  \
0  ['TFT8_Gangplank', 'TFT8_Vi', 'TFT8_LeeSin', '...   
1  ['TFT8_Lux', 'TFT8_Annie', 'TFT8_Yuumi', 'TFT8...   
2  ['TFT8_Galio', 'TFT8_Nasus', 'TFT8_Gangplank',...   
3  ['TFT8_Fiora', 'TFT8_Yasuo', 'TFT8_Vayne', 'TF...   
4  ['TFT8_Lulu', 'TFT8_Annie', 'TFT8_Yuumi', 'TFT...   
5  ['TFT8_Talon', 'TFT8_Blitzcrank', 'TFT8_Vi', '...   
6  ['TFT8_Sivir', 'TFT8_Velkoz', 'TFT8_Chogath', ...   
7  ['TFT8_Lulu', 'TFT8_Annie', 'TFT8_Yuumi', 'TFT...   
8  ['TFT8_Lulu', 'TFT8_Annie', 'TFT8_Sona', 'TFT8...   
9  ['TFT8_Rell', 'TFT8_Yuumi', 'TFT8_Nilah', 'TFT...   

          List Number Items       List Champions Tier  
0     [1, 3, 0, 2, 3, 3, 0]     [2, 3, 2, 3, 3, 3, 1]  
1  [1, 3, 1, 0, 0, 1, 3, 0]  [2, 3, 2, 2, 2, 2, 2, 1]  
2  [1, 0, 3, 0, 0, 3, 0, 1]  [2, 2, 3, 2, 2, 3, 2, 1]  
3  [0, 1, 1, 0, 1, 1, 3, 3]  [2, 2, 2, 2, 2, 2, 2, 2]  
4  [0, 3, 1, 0, 1, 3, 1, 3]  [2, 2, 2, 2, 2, 2, 2, 2]  
5  [0, 2, 0, 3, 0, 0, 1, 1]  [2, 2, 1, 3, 1, 2, 2, 2]  
6  [0, 0, 3, 0, 3, 3, 0, 1]  [2, 2, 3, 2, 2, 2, 2, 1]  
7  [0, 3, 1, 0, 3, 3, 1, 0]  [2, 2, 1, 2, 2, 1, 1, 1]  
8     [0, 3, 0, 0, 3, 1, 1]     [2, 3, 2, 1, 2, 2, 2]  
9  [3, 0, 0, 1, 0, 1, 3, 0]  [2, 2, 2, 1, 1, 2, 1, 1]

Số hàng và cột của dataframe sau khi loại bỏ các đội hình bị trùng:

In [7]:
row, col = data_team_df.shape
row, col

(18096, 8)

Ý nghĩa của từng cột:
- **Puuid**: id của các người chơi .
- **ID Match**: id của trận đấu.
- **Placement**: thứ hạng của các người chơi trong 1 trận đấu.
- **List Augments**: danh sách lõi nâng cấp trong trận đấu của 1 người chơi.
- **List Traits**: danh sách tộc, hệ được kích hoạt của 1 đội hình.
- **List Champions**: danh sách các tướng có trong 1 đội hình.

Kiểu dữ liệu của từng cột trong dataframe:

In [8]:
col_dtypes = data_team_df.dtypes
col_dtypes

Puuid                  object
ID Match               object
Placement               int64
List Augments          object
List Traits            object
List Champions         object
List Number Items      object
List Champions Tier    object
dtype: object

Dữ liệu có một số cột có kiểu dữ liệu là list nhưng mà khi đọc vào dataframe thì nó chuyển thành series và nó được hiểu là kiểu string nên ta cần chuyển nó lại dạng list

In [9]:
def changeStringToList(col_name, df):
    for i in range(len(col_name)):
        temp = []
        for j in range(len(df[col_name[i]])):
            temp.append(ast.literal_eval(df[col_name[i]][j]))
        df[col_name[i]] = temp

In [10]:
col_name_type_list = ['List Augments', 'List Champions', 'List Traits', 'List Number Items', 'List Champions Tier']
changeStringToList(col_name_type_list, data_team_df)

## Preprocess

### Kiểm tra dữ liệu bị trống

In [11]:
data_team_df.isna().sum()

Puuid                  0
ID Match               0
Placement              0
List Augments          0
List Traits            0
List Champions         0
List Number Items      0
List Champions Tier    0
dtype: int64

Không có những dữ liệu bị thiếu trong dataframe này.

### Thông tin về người chơi:

- Tổng số người chơi có trong tập dữ liệu

In [12]:
len(data_team_df['Puuid'].unique())

2449

- Tổng số trận đấu trong tập dữ liệu:

In [13]:
len(data_team_df['ID Match'].unique())

2262

Dựa trên dữ liệu này ta sẽ thấy được rằng số dữ liệu bị trùng ở trên tìm ra được là 13904, nó sẽ bằng với tổng số trận đấu trừ đi số trận bị trùng nhân với 8 (vì có 8 đội hình), vì 13904 = 32000 - 2262*8

- Người chơi có số trận đấu nhiều nhất là:

In [14]:
data_team_df['Puuid'].value_counts().max()

70

- Số lượng đội hình được sử dụng trong các trận đấu:

In [15]:
num_of_team = set()
for i in range(len(data_team_df['List Champions'])):
    num_of_team.add(str(data_team_df['List Champions'][i]))
len(num_of_team)

12530

## Khám phá dữ liệu về trang bị của tướng trong tập dữ liệu

### Đọc data từ dataframe

In [30]:
data_item_df = pd.read_csv('process_data_tft_each_item_csv.csv', usecols=['Puuid','ID Match','Champion','List Name Item'])

In [31]:
data_item_df[:10]

Puuid       ID Match  \
0  _NjJNkxpeKQ-po15A8N1GVKYqYTAjiN1WoqkVUGfwmpHqg...  KR_6267320509   
1  WgAZ0CQHGmDf8yctPx5SXkU-rt-9HWv758OmfE6t_Wod8X...  KR_6267320509   
2  WO7vH445Mvh-59sDpk4OAtffIxLMrm2R4wsrGT5ZcUb7cN...  KR_6267320509   
3  LjGpzsuqxxKX_8HDCyOmX5bBgwC-p59Oj7rTLhaSwOgN7U...  KR_6267320509   
4  XdejWcFYkQS2jJwSPR2FXlv_IZuw1txCPOLZJtUkqyusTn...  KR_6267320509   
5  CtB6YjZTon2Z-pYgNfubDLUKwxAZoBQcZVQ12LyNanULi3...  KR_6267320509   
6  y6gI4h6o614IhVvGBBp45jA-XhRHxqx2tleVdE0egluOSK...  KR_6267320509   
7  YFHBmecEZcAKDUkoNKCvM_--z5qekvafrDJzBgqJFz6l5a...  KR_6267320509   
8  y6gI4h6o614IhVvGBBp45jA-XhRHxqx2tleVdE0egluOSK...  KR_6267390363   
9  GSXdu-IDqnhxAjE9OC-rEbktUy7Q-YfRmdz7P6nztlNYHj...  KR_6267390363   

                                            Champion  \
0  ['TFT8_Gangplank', 'TFT8_Vi', 'TFT8_LeeSin', '...   
1  ['TFT8_Lux', 'TFT8_Annie', 'TFT8_Yuumi', 'TFT8...   
2  ['TFT8_Galio', 'TFT8_Nasus', 'TFT8_Gangplank',...   
3  ['TFT8_Fiora', 'TFT8_Yasuo', 'TFT8_Vayne', 'TF...   
4  ['TFT8_Lulu', 'TFT8_Annie', 'TFT8_Yuumi', 'TFT...   
5  ['TFT8_Talon', 'TFT8_Blitzcrank', 'TFT8_Vi', '...   
6  ['TFT8_Sivir', 'TFT8_Velkoz', 'TFT8_Chogath', ...   
7  ['TFT8_Lulu', 'TFT8_Annie', 'TFT8_Yuumi', 'TFT...   
8  ['TFT8_Lulu', 'TFT8_Annie', 'TFT8_Sona', 'TFT8...   
9  ['TFT8_Rell', 'TFT8_Yuumi', 'TFT8_Nilah', 'TFT...   

                                      List Name Item  
0  [['TFT_Item_ZekesHerald'], ['TFT_Item_BrambleV...  
1  [['TFT8_Item_HeartEmblemItem'], ['TFT_Item_Bra...  
2  [['TFT_Item_GargoyleStoneplate'], [], ['TFT_It...  
3  [[], ['TFT8_Item_HackerEmblemItem'], ['TFT_Ite...  
4  [[], ['TFT_Item_WarmogsArmor', 'TFT_Item_Titan...  
5  [[], ['TFT_Item_TitanicHydra', 'TFT_Item_Titan...  
6  [[], [], ['TFT_Item_TitansResolve', 'TFT_Item_...  
7  [[], ['TFT_Item_WarmogsArmor', 'TFT_Item_Gargo...  
8  [[], ['TFT_Item_GargoyleStoneplate', 'TFT_Item...  
9  [['TFT_Item_IonicSpark', 'TFT_Item_Redemption'...

## Data exploration

Dữ liệu có bao nhiêu hàng và cột:

In [32]:
row, col = data_item_df.shape
row, col

(32000, 4)

In [33]:
have_duplicated_rows = data_item_df.duplicated('List Name Item').sum()
have_duplicated_rows

13909

Số lượng dữ liệu trùng cũng giống bên dữ liệu đội hình.

In [34]:
data_item_df = data_item_df.drop_duplicates().reset_index(drop = True)

Ý nghĩa của từng cột:
- **Puuid**: id của các người chơi .
- **ID Match**: id của trận đấu.
- **Champion**: danh sách tướng trong 1 đội hình.
- **List Augments**: danh sách trang bị ứng với từng tướng trong đội hình.

Kiểu dữ liệu của từng cột trong dataframe

In [35]:
col_dtypes = data_item_df.dtypes
col_dtypes

Puuid             object
ID Match          object
Champion          object
List Name Item    object
dtype: object

Có 2 cột dữ liệu dạng list nhưng khi lưu vào csv thì nó chuyển thành string nên ta cần chuyển nó trở lại về list

In [36]:
col_name_type_list = ['Champion', 'List Name Item']
changeStringToList(col_name_type_list, data_item_df)

## Preprocess

### Kiểm tra dữ liệu trống:

In [37]:
data_item_df.isna().sum()

Puuid             0
ID Match          0
Champion          0
List Name Item    0
dtype: int64

Không có dữ liệu bị trống trong dataframe này

# Thông tin về tướng và trang bị

- Tổng số tướng được sử dụng:

In [38]:
a = np.array(data_item_df['Champion'])
numbers_champ = set(i for j in a for i in j)
print(len(numbers_champ)) # 10

59


- Tổng số item được sử dụng

In [42]:
b = np.array(data_item_df['List Name Item'])
numbers_item = set()
for i in b:
    for j in i:
        for k in j:
            numbers_item.add(k)

len(numbers_item)

124